catalog/async/catalog-async-data/src/main/java/org/codice/ddf/catalog/async/data/impl/LazyProcessResourceImpl.java

```diff
@@ -0,0 +1,171 @@
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.function.Supplier;
+
+import org.apache.commons.io.IOUtils;
+import org.apache.commons.lang.StringUtils;
+import org.codice.ddf.catalog.async.data.api.internal.ProcessResource;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import ddf.catalog.resource.Resource;
+
+public class LazyProcessResourceImpl implements ProcessResource {
+    private static final Logger LOGGER = LoggerFactory.getLogger(LazyProcessResourceImpl.class);
+
+    private URI uri;
+
+    private String name;
+
+    private String mimeType;
+
+    private long size;
+
+    private InputStream inputStream;
+
+    private String qualifier;
+
+    private boolean isModified = false;
+
+    private boolean isResourceLoaded = false;
+
+    private Supplier<Resource> resourceSupplier;
+
+    private String metacardId;
+
+    public LazyProcessResourceImpl(String metacardId, Supplier<Resource> resourceSupplier) {
+        if (StringUtils.isBlank(metacardId)) {
+            throw new IllegalArgumentException(
+                    "ProcessResourceImpl argument \"metacardId\" may not be null or empty.");
+        }
+        if (resourceSupplier == null) {
+            throw new IllegalArgumentException(
+                    "LazyProcessResourceImpl must have a non null resource supplier");
+        }
+
+        this.isModified = false;
+        this.metacardId = metacardId;
+        this.resourceSupplier = resourceSupplier;
+        this.isResourceLoaded = false;
+    }
+
+    @Override
+    public URI getUri() {
+        loadResource();
+        return uri;
+    }
+
+    @Override
+    public String getName() {
+        loadResource();
+        return name;
+    }
+
+    @Override
+    public String getQualifier() {
+        loadResource();
+        return qualifier;
+    }
+
+    @Override
+    public String getMimeType() {
+        loadResource();
+        return mimeType;
+    }
+
+    @Override
+    public InputStream getInputStream() throws IOException {
+        loadResource();
+        return inputStream;
+    }
+
+    @Override
+    public long getSize() {
+        loadResource();
+        return size;
+    }
+
+    @Override
+    public boolean isModified() {
+        return isModified;
+    }
+
+    @Override
+    public void close() {
+        IOUtils.closeQuietly(inputStream);
+    }
+
+    public void markAsModified() {
+        isModified = true;
+    }
+
+    private void loadResource() {
+        if (!isResourceLoaded) {
+            populateProcessResource(resourceSupplier.get());
+        }
+    }
+
+    private void populateProcessResource(Resource resource) {
+        if (resource == null) {
+            LOGGER.debug(
+                    "Error loading resource for metacard id: {}. null resource was returned by supplier.",
+                    metacardId);
+            return;
```

Comment:
✏️ If we add `loadResource()` to the `ProcessResource` interface we can avoid having callers having to check the value of the getters to ensure the resource was loaded, since we just return here.

Reply:
Not adding loadResource to the interface